Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Win32Application.cpp

```diff
@@ -0,0 +1,164 @@
+#include "pch.h"
+#include "stdafx.h"
+#include "Win32Application.h"
+
+HWND Win32Application::m_hwnd = nullptr;
+bool Win32Application::close_window = false;
+
+// Get the HModule that will be used to spawn
+// the HWND
+HMODULE GetCurrentModule()
+{
+    HMODULE hModule = NULL;
+    GetModuleHandleEx(
+        GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS,
+        (LPCTSTR)GetCurrentModule,
+        &hModule);
+    return hModule;
+}
+
+// Launch the HWND and init D3D
+int Win32Application::Run(D3D12Quad* pSample, int nCmdShow)
+{
+    // Get the Hinstance
+    HINSTANCE hInstance = GetCurrentModule();
+
+    // Initialize the window class.
+    WNDCLASSEX windowClass = { 0 };
+    windowClass.cbSize = sizeof(WNDCLASSEX);
+    windowClass.style = CS_HREDRAW | CS_VREDRAW;
+    windowClass.lpfnWndProc = WindowProc;
+    windowClass.hInstance = hInstance;
+    windowClass.hCursor = LoadCursor(NULL, IDC_ARROW);
+    windowClass.lpszClassName = L"DXQuad";
+
+    auto hr = RegisterClassEx(&windowClass);
+    // Register window class
+    auto err = GetLastError();
+    if (!hr)
+    {
+        MessageBox(NULL, L"Error registering class",
+            L"Error", MB_OK | MB_ICONERROR);
+        return false;
+    }
+
+    RECT windowRect = { 0, 0, static_cast<LONG>(pSample->GetWidth()), static_cast<LONG>(pSample->GetHeight()) };
+    AdjustWindowRect(&windowRect, WS_OVERLAPPEDWINDOW, FALSE);
+
+    // Create the window and store a handle to it.
+    m_hwnd = CreateWindow(
+        windowClass.lpszClassName,
+        pSample->GetTitle(),
+        WS_OVERLAPPEDWINDOW,
+        CW_USEDEFAULT,
+        CW_USEDEFAULT,
+        windowRect.right - windowRect.left,
+        windowRect.bottom - windowRect.top,
+        nullptr,        // We have no parent window.
+        nullptr,        // We aren't using menus.
+        hInstance,
+        pSample);
+    if (!m_hwnd) { MessageBox(NULL, L"Error creating window", L"Error", MB_OK | MB_ICONERROR); return false; }
+
+    // Initialize the sample and show the window
+    pSample->OnInit();
+    ShowWindow(m_hwnd, nCmdShow);
+
+    // Main sample loop.
+    close_window = false;
+    MSG msg = {};
+    while (msg.message != WM_QUIT)
+    {
+        // Destroy the sample and window if the close_window
+        // variable is set to true
+        if (close_window)
+        {
+            pSample->OnDestroy();
+            PostMessage(m_hwnd, WM_CLOSE, 0, 0);
+            break;
+        }
+        // Process any messages in the queue.
+        if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE))
```

Comment:
Does this app repeatedly draw frames with changing content? If not, `GetMessage` is the correct one here because `PeekMessage` returns immediately and burns idle CPU cycles (https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage), whereas GetMessage waits until a mouse message or keypress or other significant event.

```while (GetMessage() > 0)
{
    ...
}
```